### PR TITLE
Add minishell checker

### PR DESCRIPTION
**Pull request by gabrielmocanu**
Add minishell checker

### CODE REVIEW

**Comment by gabrielmocanu** [comment]
@teodutu 

This is a draft, I had to resolve the linting for sure in the script, but you can review it anyways.
I don't think it will be new files for this PR.

File: content/assignments/minishell/checker/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+```console
+student@so:~/.../assignments/minishell/checker/_test/inputs$ ls -F
+test_01.txt  test_03.txt  test_05.txt  test_07.txt  test_09.txt  test_11.txt  test_13.txt  test_15.txt  test_17.txt
+test_02.txt  test_04.txt  test_06.txt  test_08.txt  test_10.txt  test_12.txt  test_14.txt  test_16.txt  test_18.txt
+```
+
+To execute tests you need to run:
+
+```bash
```
Comment (teodutu) [inline content/assignments/minishell/checker/README.md:14]: Change this hint to `console`.

File: content/assignments/minishell/checker/README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+./run_all.sh
```
Comment (teodutu) [inline content/assignments/minishell/checker/README.md:15]:
```suggestion
student@os:~/.../assignments/minishell/checker$ ./run_all.sh
```
And add a sample output that gets full marks.

File: content/assignments/minishell/checker/README.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
+# Testing and Grading
+
+The testing is automated.
+Tests are located in the `inputs/` directory.
+
+```console
+student@so:~/.../assignments/minishell/checker/_test/inputs$ ls -F
```
Comment (teodutu) [inline content/assignments/minishell/checker/README.md:7]:
```suggestion
student@os:~/.../assignments/minishell/checker/_test/inputs$ ls -F
```

File: content/assignments/minishell/checker/README.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+## Memory leaks
+
+To inspect the un-released resources (memory leaks, file descriptors) set `USE_VALGRIND=yes` and `DO_CLEANUP=no` in `_test/run_test.sh`.
+You can modify both the path to the valgrind log file and the command parameters.
+To see the results of the tests, you can check `_test/outputs/` directory.
+
+## Checkpatch
+
+TBD
```
Comment (teodutu) [inline content/assignments/minishell/checker/README.md:31]: You can instruct students on how to download and run the script.

File: content/assignments/minishell/checker/_test/inputs/test_15.txt [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+echo '#include <stdlib.h>' >> main.c
+echo 'int main(int argn, char *argv[]) {' >> main.c
+echo 'FILE *g=fopen(argv[1],"w");' >> main.c
+echo 'int i;' >> main.c
+echo 'for(i = 0 ; i< 10 * 1024 * 1024 && fprintf(g,"*");i++);' >> main.c
+echo 'return 0; }' >> main.c
+gcc -o main.exe main.c
+./main.exe big_file
+rm -fr main.exe
+cat < big_file | cat | cat | cat | cat | cat | cat > firstFIle & cat < big_file | cat | cat | cat | cat | cat | cat > secondFIle
```
Comment (teodutu) [inline content/assignments/minishell/checker/_test/inputs/test_15.txt:12]: :laughing:

**Review by teodutu — CHANGES_REQUESTED** [review]
This looks mostly OK. Have you considered renaming the test files according to what they test? I guess it might be easier for students to see what's failing when debugging.

File: content/assignments/minishell/checker/README.md (hunk 1 above)
Comment (gabrielmocanu) [reply content/assignments/minishell/checker/README.md:14]: Done since README is no longer in this folder.
It's already in the PR with the assignment.

File: content/assignments/minishell/checker/README.md (hunk 2 above)
Comment (gabrielmocanu) [reply content/assignments/minishell/checker/README.md:15]: Done since README is no longer in this folder.
It's already in the PR with the assignment.
Add this review to this [PR] if you consider this useful.
[PR]: https://github.com/open-education-hub/operating-systems/pull/111

File: content/assignments/minishell/checker/README.md (hunk 3 above)
Comment (gabrielmocanu) [reply content/assignments/minishell/checker/README.md:7]: Done since README is no longer in this folder.
It's already in the PR with the assignment.

File: content/assignments/minishell/checker/README.md (hunk 4 above)
Comment (gabrielmocanu) [reply content/assignments/minishell/checker/README.md:31]: Done since README is no longer in this folder.
It's already in the PR with the assignment.

File: content/assignments/minishell/checker/_test/inputs/test_15.txt (hunk 5 above)
Comment (gabrielmocanu) [reply content/assignments/minishell/checker/_test/inputs/test_15.txt:12]: Done since README is no longer in this folder.
It's already in the PR with the assignment.

**Comment by gabrielmocanu** [comment]
> This looks mostly OK. Have you considered renaming the test files according to what they test? I guess it might be easier for students to see what's failing when debugging.

This would take a complete rewrite for the checker.
I don't think this is a big problem, because they have the test index also in the checker output display.

**Comment by gabrielmocanu** [comment]
@teodutu 

Please check again.
FMPOV all the errors from the checkpatch and superlinter are not relevant.

File: content/assignments/minishell/checker/_test/run_test.sh [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+	test_common		"Testing multiple pipes"		10	\
+	test_common		"Testing variables and redirect"	5	\
+	test_common		"Testing overwritten variables"		2	\
+	test_common		"Testing all operators"			2	\
+	test_common_alt		"Testing parallel operator"		10	\
+	test_common_alt		"Testing big file"			5	\
+	test_common_alt		"Testing sleep command"			7	\
+	test_common_alt		"Testing fscanf function"		7	\
+	test_exec_failed	"Testing unknown command"		4	\
+)
```
Comment (teodutu) [inline content/assignments/minishell/checker/_test/run_test.sh:239]: The total score is 95. For the previous assignments, the total was 90. I think we should stick to that.

File: content/assignments/minishell/checker/_test/inputs/test_18.txt [hunk 7]
```diff
@@ -0,0 +1,5 @@
+it_doesnt_exist
+true && it_doesnt_exist
+it_doesnt_exist && echo "test"
+executabil_care_nu_exista || echo "test"
```
Comment (teodutu) [inline content/assignments/minishell/checker/_test/inputs/test_18.txt:4]:
```suggestion
executable_that_does_not_exist || echo "test"
```

File: content/assignments/minishell/checker/_test/run_test.sh [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+INPUT_DIR="_test/inputs"
+REFS_DIR="_test/refs"
+LOG_FILE="/dev/null"
+export max_points=95
+TEST_TIMEOUT=30
+
+TEST_LIB=_test/test_lib.sh
+export CHECK_SKIP='parser.\(\(tab\|yy\).\(c\|h\)\|h\)$'
+
+# load the lib functions
```
Comment (teodutu) [inline content/assignments/minishell/checker/_test/run_test.sh:20]:
```suggestion
# Load the lib functions.
```
Change all comments to start with capital letters and end with a dot (`.`).

**Review by teodutu — CHANGES_REQUESTED** [review]
I agree with the marks given for each test, just that their sum should probably add up to 90, not 95.

File: content/assignments/minishell/checker/_test/run_test.sh (hunk 8 above)
Comment (gabrielmocanu) [reply content/assignments/minishell/checker/_test/run_test.sh:20]: Done.

File: content/assignments/minishell/checker/_test/inputs/test_18.txt (hunk 7 above)
Comment (gabrielmocanu) [reply content/assignments/minishell/checker/_test/inputs/test_18.txt]: Done.

File: content/assignments/minishell/checker/_test/run_test.sh [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+	OUT_DIR="${TEST_NAME}_out"
+	REF_DIR="${TEST_NAME}_ref"
+	IN_FILE="${TEST_NAME}.in"
+	REF_FILE="${TEST_NAME}.ref"
+	OUT_FILE="${TEST_NAME}.out"
+
+	local input_basename="${INPUT_DIR}/${TEST_NAME}"
+	[ -f "${input_basename}.${OS_PLATFORM}.txt" ] && \
+		input_basename="${input_basename}.${OS_PLATFORM}" || \
+		input_basename="${input_basename}"
```
Comment (darius-m) [inline content/assignments/minishell/checker/_test/run_test.sh:77]: Add this as an `if` statement, instead of ternary-like code. The `[ cond ] && if_true || if_false` is only readable if everything is on the same line, otherwise the `&&` and `||` are harder to spot.

File: content/assignments/minishell/checker/_test/run_test.sh [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+# Change this of you want to keep the logs after execution.
+DO_CLEANUP=no
+
+BUFFERING_WRAPPER=""
+if command -v stdbuf &> /dev/null; then
+	# stdbuf may not be installed on cygwin.
+	BUFFERING_WRAPPER="stdbuf -i 0"
+fi
+
+USE_VALGRIND=yes
```
Comment (darius-m) [inline content/assignments/minishell/checker/_test/run_test.sh:38]: Enabling or disabling valgrind may be easier if the variable is checked and assigned a default value if not defined in the environment - e.g. `USE_VALGRIND="${USE_VALGRIND:-yes}"`.

File: content/assignments/minishell/checker/_test/run_test.sh [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+{
+
+	if ! [ -e "$exec_name" ]; then
+		echo "$exec_name not found! Cannot run the test"
+		exit 1
+	fi
+
+	if [ "$USE_VALGRIND" = "yes" ]; then
+		which valgrind &> /dev/null
+		if [ $? -eq 0 ]; then
```
Comment (darius-m) [inline content/assignments/minishell/checker/_test/run_test.sh:54]: Simply use the result of the command, instead of checking the return code separately.
```suggestion
		if which valgrind &> /dev/null; then
```

Also, for the `stdbuf` check above, the `command` builtin is used. It may be worth using only one of them across the entire script.

File: content/assignments/minishell/checker/_test/run_test.sh [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+	REF_FILE="${TEST_NAME}.ref"
+	OUT_FILE="${TEST_NAME}.out"
+
+	local input_basename="${INPUT_DIR}/${TEST_NAME}"
+	[ -f "${input_basename}.${OS_PLATFORM}.txt" ] && \
+		input_basename="${input_basename}.${OS_PLATFORM}" || \
+		input_basename="${input_basename}"
+
+	# Copy input files in the testing directory.
+	cp "${input_basename}.txt" "${MAIN_TEST_DIR}/${IN_FILE}" &> "$LOG_FILE"
```
Comment (darius-m) [inline content/assignments/minishell/checker/_test/run_test.sh:80]: `cp` will print nothing by default (i.e. `-v` is not used), so the redirect does nothing. Also, if the `LOG_FILE` variable were a proper file, it would be emptied by the non-appending `>` redirect.

File: content/assignments/minishell/checker/_test/run_test.sh [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+	# cp ${input_basename}?*.txt ${MAIN_TEST_DIR} &> $LOG_FILE
+	# Copy any reference files, if they exist (see test 18).
+	cp "${REFS_DIR}/${TEST_NAME}.ref" "${MAIN_TEST_DIR}" &> "$LOG_FILE"
+
+	# Export mini-shell in path.
+	PATH="$PATH:$(pwd)"
+	cd "${MAIN_TEST_DIR}" &> "$LOG_FILE" || exit 1
+
+	[ -d "${OUT_DIR}" ] && rm -rf "${OUT_DIR}"
+	[ -d "${REF_DIR}" ] && rm -rf "${REF_DIR}"
```
Comment (darius-m) [inline content/assignments/minishell/checker/_test/run_test.sh:91]: Checking if the directories exist is not needed, as the `-f` flag should silence missing files warnings.

File: content/assignments/minishell/checker/_test/run_test.sh [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+	# target implementation with the reference output starts before the
+	# target implementation has finished writing files to disk. Handle this
+	# by mandating that all I/O buffers be flushed.
+	sync
+	cd - &> "$LOG_FILE" || exit 1
+}
+
+
+generate_input()
+{
```
Comment (darius-m) [inline content/assignments/minishell/checker/_test/run_test.sh:142]: It's hard to read where the variables are set if they are not passed as function parameters.

File: content/assignments/minishell/checker/_test/run_test.sh [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+
+# Checks the output of the commands.
+test_output()
+{
+	init_test
+
+	generate_input
+
+	# Commands to execute the test.
+	execute_cmd "$ref_name" "../${REF_FILE}" "${OUT_FILE}"
```
Comment (darius-m) [inline content/assignments/minishell/checker/_test/run_test.sh:155]: It's hard to follow why the `../${REF_FILE}` is needed here, but not in `generate_input`. I assume it's because of the `cd` in `execute_cmd`, but I'm not sure. It may be worth keeping the absolute path to the files (e.g. using `local ref=$(readlink -e "${REF_FILE})` and passing that as the argument.

File: content/assignments/minishell/checker/_test/run_test.sh [hunk 16]
```diff
@@ -0,0 +1,244 @@
...
+}
+
+# ----------------- Test Suite ----------------------------------------------- #
+
+# Specific test example.
+execute_cmd()
+{
+	EXEC=$1
+	INPUT=$2
+	OUTPUT=$3
```
Comment (darius-m) [inline content/assignments/minishell/checker/_test/run_test.sh:128]: The arguments should normally be escaped. For the first argument (`EXEC`) this may be an issue since you need having multiple arguments when running `valgrind`. I think it would be easier to have a `GEN_REF` flag (or similar) and use a function to run the correct command - e.g.,
```bash
eval_command() {
    GEN_REF="$1"
    shift 1 # drop the first argument
    if [ "${GEN_REF}" = "yes" ]; then
       "$ref_name"
    elif [ "${USE_VALGRIND}" = "yes" ]; then
       # valgrind and all required arguments
    else
       "${exec_name}"
    fi
}
```

**Review by darius-m — COMMENTED** [review]
I have added some recommendations for the `run_test.sh` script, mostly to improve readability. Simply mark them as solved if no action is intended.

File: content/assignments/minishell/checker/_test/run_test.sh (hunk 6 above)
Comment (teodutu) [reply content/assignments/minishell/checker/_test/run_test.sh]: IIRC, the discussion was to award the last 10 points (from 90 to 100) after checking the assignment manually. This means going beyond checkpatch and checking for code coherence, modularity etc.

Comment (gabrielmocanu) [reply content/assignments/minishell/checker/_test/run_test.sh]: Yes, you are right. But I don't understand your suggestion.
How do you want to handle this?
Exclude checkpatch from checker? 
Let it to run only with a special flag without any scoring?
Adjust the score to 90/100 and still exclude checkpatch? Adjust the score to 90/90?

Comment (teodutu) [reply content/assignments/minishell/checker/_test/run_test.sh]: Keep checkpatch in the checker, but lower the score to 90/100, like the other assignments.

File: content/assignments/minishell/checker/_test/run_test.sh [hunk 17]
```diff
@@ -0,0 +1,244 @@
...
+	test_common		"Testing multiple pipes"		10	\
+	test_common		"Testing variables and redirect"	5	\
+	test_common		"Testing overwritten variables"		2	\
+	test_common		"Testing all operators"			2	\
+	test_common_alt		"Testing parallel operator"		10	\
+	test_common_alt		"Testing big file"			5	\
+	test_common_alt		"Testing sleep command"			7	\
+	test_common_alt		"Testing fscanf function"		7	\
+	test_exec_failed	"Testing unknown command"		4	\
+)
```
Comment (teodutu) [inline content/assignments/minishell/checker/_test/run_test.sh:234]: Now they add up to 100. Make them add up to 90/100 and we'll add the remaining 10 points manually.

**Review by teodutu — CHANGES_REQUESTED** [review]
I've re-explained what I meant regarding the test scores. Also mind the suggestions from @darius-m.

**Comment by teodutu** [comment]
@gabrielmocanu bump

File: content/assignments/minishell/checker/_test/run_test.sh [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+	# by mandating that all I/O buffers be flushed.
+	sync
+	cd - &> "$LOG_FILE" || exit 1
+}
+
+
+generate_input()
+{
+	# Generate the PROMPT for BASH.
+	sed 's/\(.*\)/printf \"> \"; \1/' "${IN_FILE}" > "${REF_FILE}"
```
[the remark this answers is not in the document]
Comment (gabrielmocanu) [reply content/assignments/minishell/checker/_test/run_test.sh:144]: It doesn't work with the 2nd approach but I don't understand why.
I will let it as it is because it's not a big difference and I don't want to keep this PR open.
We can review this after with another PR / issue.

File: content/assignments/minishell/checker/_test/run_test.sh (hunk 15 above)
Comment (gabrielmocanu) [reply content/assignments/minishell/checker/_test/run_test.sh:155]: Agree.
I will make another issue and solve it with another PR.

**Comment by gabrielmocanu** [comment]
@darius-m 
Thanks for the review. I applied most of your comments.
For 1 or 2 I will create issues and resolve them after because are more stylistic issues.

**Comment by gabrielmocanu** [comment]
@teodutu 
All issues from `checkpatch` and `super-linter` are false positive FMPOV.
If you want to address anything from there just let me know and create an issue for this.
Meanwhile, I will merge this PR.
